### ISSUE TITLE
Implement `starknet_getStateUpdate`

### ISSUE THREAD

**Issue by tarrencev** [eqlabs/pathfinder#417]
Starknet now supports contract factory patterns and it is currently not possible to index a contract deployed using the deploy `syscall`.

**Comment by Mirko-von-Leipzig** [comment]
I think an issue is that pathfinder can't validate these traces locally without re-running all transactions again locally?

I'm very much a noob on these topics though, so maybe I'm misremembering/misunderstanding this.

**Comment by xJonathanLEI** [comment]
Am pretty sure `pathfinder` can't validate the traces. IMO `pathfinder` should never serve data it can't validate. So tracing APIs should not be implemented at this point. But that's just me.

**Comment by xJonathanLEI** [comment]
That said, I believe pathfinder is _already_ serving data it cannot validate? Notably the transaction receipts, which I don't believe is part of the proof? I could be wrong though.

**Comment by Mirko-von-Leipzig** [comment]
> That said, I believe pathfinder is already serving data it cannot validate? Notably the transaction receipts

That is true; and technically a whole lot more currently. Technically, only the state root is on L1 (and not the block hash... :( ) which means that only the state updates i.e. contracts, and storage updates are verifiable.

So no block head / transaction / transaction receipts are verifiable against L1. If the block hash were included on L1 we could verify these as well -- including transaction receipts (well sort of -- not the content, just the count I think?). afaik this is supposed to be possible at some point in the future.

**Comment by xJonathanLEI** [comment]
I understand the whole point of L2 is that such re-execution should not be necessary, but that doesn't mean we can't do it, especially when we want data that are not part of the proof.

**Comment by xJonathanLEI** [comment]
But back to the issue of indexing contract deployments itself, I think it's sufficient to have the state update API implemented instead of the trace API. Implementing the state update API should not suffer from the same issue.

**Comment by xJonathanLEI** [comment]
> This is not like a strong proof of correctness since you can technically have different transactions resulting the same state diff

Taking a step back, given the lack of consensus mechanism on StarkNet (not even PoA), how do you even argue it's not a valid block if it fulfills the state diffs? All blocks that fullfill the diffs are valid blocks IMO.

**Comment by Mirko-von-Leipzig** [comment]
> All blocks that fullfill the diffs are valid blocks IMO.

This is true. Which is why a block hash or transactions need to be included in a proof. Re-running transactions locally sounds awful given how slow `call` itself currently is.

**Comment by xJonathanLEI** [comment]
Well running transactions locally can be done in parallel between blocks, so it shouldn't be that bad. I personally would definitely run one given all the errors in sequencer receipt data we've seen so far.

**Comment by Mirko-von-Leipzig** [comment]
For new blocks on mainnet that's fine. But doing so for sync'ing is prohibitive, at least with how linearly setup our sync process currently is.

Also not sure it would be fast enough for live Goerli. Given that a single call takes several seconds at least and that they cannot execute in parallel.

**Comment by xJonathanLEI** [comment]
Yeah but the sequencer will eventually be decentralized, at which point it will likely no longer make sense to trust it for receipts & traces.

**Comment by xJonathanLEI** [comment]
As mentioned I believe for indexing contracts it should be sufficient to have the state update method. @tarrencev Could you confirm?

**Comment by tarrencev** [comment]
@xJonathanLEI  yup that's correct. would be great to expose declared contracts in the state diff

**Comment by Mirko-von-Leipzig** [comment]
We will work on supporting `starknet_getStateUpdate` 👍

**Comment by Mirko-von-Leipzig** [comment]
Implemented in #484. Note that this does not include an automatic migration to fix the data missing for blocks already in the database - this would require downloading them all again as the data simply isn't present currently.

We will release a new testnet snapshot which will include this state update data when we release the next pathfinder version early next week.
